### PR TITLE
feat: spice: implement data distribution for mvp

### PR DESCRIPTION
**Pull request by attila0x2A**
This PR implements spice data distribution logic assuming all nodes are always online and available (i.e. using distribution based on pushing data only) without banning or de-prioritization of malicious nodes.

Since in practice execution is likely to lag behind block production and all data distributed with spice data distribution will be based on execution, I decided to have execution wait on block corresponding to the data received before processing it while having receipt verification logic in chunk executor actor to avoid having distributor wait on execution results and keep less message-specific logic in it.

Part of https://github.com/near/nearcore/issues/13611

### CODE REVIEW

File: chain/client/src/chunk_executor_actor.rs [hunk 1]
```diff
@@ -708,19 +737,86 @@ impl ChunkExecutorActor {
...
+    }
+
+    fn process_new_receipts(
+        &mut self,
+        receipts: ExecutorIncomingUnverifiedReceipts,
+    ) -> Result<(), Error> {
+        let block_hash = receipts.block_hash;
+
+        let verified_receipts = match self.receipts_verification_context(&block_hash)? {
+            ReceiptVerificationContext::Ready { execution_results } => {
```
Comment (pugachAG) [inline chain/client/src/chunk_executor_actor.rs]: nit: as far as I understand we don't have to perform verification here and just save receipts to the pending queue, so this is looks like an optimization maybe we'd better avoid for now to reduce code duplication, up to you

File: chain/client/src/tests/spice_data_distributor_actor.rs [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+    with_10_producers_1000_validators(10, 1000)
+    with_100_producers_1_validator(100, 1)
+    with_100_producers_10_validators(100, 10)
+    with_100_producers_100_validators(100, 100)
+    with_100_producers_1000_validators(100, 1000)
+    with_5000_producers_1_validator(5000, 1)
+    with_5000_producers_10_validators(5000, 10)
+    with_5000_producers_100_validators(5000, 100)
+    with_5000_producers_1000_validators(5000, 1000)
+    with_5000_producers_4000_validators(5000, 4000)
```
Comment (pugachAG) [inline chain/client/src/tests/spice_data_distributor_actor.rs:419]: why do we need all those test cases? what is the expected difference between `10_producers_100_validators` and `100_producers_1000_validators`?

File: chain/client/src/tests/spice_data_distributor_actor.rs [hunk 3]
```diff
@@ -0,0 +1,1056 @@
...
+test_receipts_distribution! {
+    with_2_producers(2)
+    with_5_producers(5)
+    with_10_producers(10)
+    with_20_producers(20)
+    with_50_producers(50)
+    with_100_producers(100)
+    with_500_producers(500)
+    with_5000_producers(5000)
+}
```
Comment (pugachAG) [inline chain/client/src/tests/spice_data_distributor_actor.rs:544]: same here

**Review by pugachAG — APPROVED** [review]

File: chain/client/src/tests/spice_data_distributor_actor.rs (hunk 2 above)
Comment (attila0x2A) [reply chain/client/src/tests/spice_data_distributor_actor.rs:419]: Mostly to cover any possible corner cases, though yes, I may have went a bit overboard with the amount of them.

File: chain/client/src/chunk_executor_actor.rs (hunk 1 above)
Comment (attila0x2A) [reply chain/client/src/chunk_executor_actor.rs]: Good point, removed in [64f29a8](https://github.com/near/nearcore/pull/14227/commits/64f29a838aa84fcee68efa822c49f9387d632bfc)

**Comment by attila0x2A** [comment]
@saketh-are I will try merging this PR. If you have any feedback still let me know and I will address it in separate PRs.
